rfd/0169-auto-updates-linux-agents.md

```diff
@@ -0,0 +1,410 @@
+8. The rollout plan for the new agent version is not fully-configurable using tctl.
+9. Agent installation logic is spread between the auto-updater script, install script, auto-discovery script, and documentation.
+10. Teleport contains logic that is specific to Teleport Cloud upgrade workflows.
+11. The existing auto-updater is not self-updating.
+12. It is difficult and undocumented to automate agent upgrades with custom automation (e.g., with JamF). 
+
+We must provide a seamless, hands-off experience for auto-updates that is easy to maintain.
+
+## Details
+
+We will ship a new auto-updater package written in Go that does not interface with the system package manager.
+It will be distributed as a separate package from Teleport, and manage the installation of the correct Teleport agent version manually. 
+It will read the unauthenticated `/v1/webapi/find` endpoint from the Teleport proxy, parse new fields on that endpoint, and install the specified agent version according to the specified upgrade plan.
+It will download the correct version of Teleport as a tarball, unpack it in `/var/lib/teleport`, and ensure it is symlinked from `/usr/local/bin`.
+
+Source code for the updater will live in `integrations/updater`.
+
+### Installation
+
+```shell
+$ apt-get install teleport-ent-updater
+$ teleport-update enable --proxy example.teleport.sh
+
+# if not enabled already, configure teleport and:
+$ systemctl enable teleport
+```
+
+### API
+
+#### Endpoints
+
+`/v1/webapi/find`
+```json
+{
+  "server_edition": "enterprise",
+  "agent_version": "15.1.1",
+  "agent_auto_update": true,
+  "agent_update_jitter_seconds": 10
+}
+```
+Notes:
+- The Teleport proxy translates upgrade hours (below) into a specific time after which the served `agent_version` changes, resulting in all agents being upgraded.
+- Critical updates are achieved by serving the desired `agent_version` immediately.
+- If an agent misses an upgrade window, it will always update immediately due to the new agent version being served.
+- The edition served is the cluster edition (enterprise, enterprise-fips, or oss), and cannot be configured.
+
+#### Teleport Resources
+
+```yaml
+kind: cluster_maintenance_config
+spec:
+  # agent_auto_update allows turning agent updates on or off at the
+  # cluster level. Only turn agent automatic updates off if self-managed
+  # agent updates are in place.
+  agent_auto_update: true|false
+  # agent_update_hour sets the hour in UTC at which clients should update their agents.
+  agent_update_hour: 0-23
+  # agent_update_now overrides agent_update_hour and serves the new version immediately.
+  # This is useful for rolling out critical security updates and bug fixes.
+  agent_update_now: on|off
+  # agent_update_jitter_seconds sets a duration in which the upgrade will occur after the hour.
+  # The agent upgrader will pick a random time within this duration to wait to upgrade.
+  agent_update_jitter_seconds: 0-3600
+  
+  [...]
+```
```

Comment:
I might be misunderstanding something here, but the interaction between the resources (`cluster_maintenance_config` and `autoupdate_version`) the proxy, and the updater seems a little more complex than needed.

Two examples:

If a critical release needs to be pushed out, it's clear that the following needs to be set `agent_update_now=on`, but it's less clear when this can be turned off. After some period that we're happy enough agents have flipped over? When the next release comes out? What happens if we forget?

The proxy is using multiple resources to compute the time the `agent_version` field served from the `/v1/webapi/find` will change. Can we make this dumber by pushing the following information `(version, critical)`, `agent_update_hour`, `agent_update_jitter_seconds` to the agent? It's much simpler and less chances of things going wrong.

```
if ver.IsCritical() {
  // upgrade now
} else {
  if inWindow() && upgradeNeeded() {
     jitter()
     // upgrade now
  }
}
```

Reply:
After chatting with @hugoShaka, we decided to simplify things by removing the window concept from the agent entirely. This also protects against agents with clock skew, misconfigured timezones, etc. It also simplifies the updater logic and ensures that future changes to the logic can be deployed server-side without needing updates to the updater.

The agent logic is just:
```
if upgradeNeeded() {
  jitter()
  upgrade()
}
```

With this new design, when `agent_update_hour` is hit or `agent_update_now` is true, the proxies advertise the new version in `autoupdate_version`, and agents update immediately (after jitter). The `cluster_maintenance_config` fields delay the state change from whatever the version is being advertised now, to whatever version is set in `autoupdate_version`. After the new version is advertised, the `cluster_maintenance_config` fields may be adjusted to configure the behavior for the next version.

I agree that this introduces an undesirable form of statefulness, where there is a hidden "current version" that is only exposed to the user if they examine the proxy endpoint. If `agent_update_hour` were a Kubernetes resource, I would put the current version and calculated transition time in `/status`. Not sure if Teleport resources have a similar pattern for converting edge-triggered behavior to level-triggered behavior.

It occurs to me that `agent_update_now` should likely move into `autoupdate_version`, as Cloud would likely want to set the version as critical -- not the end user. It might also be renamed to `agent_version_critical`. That would make it clear that criticalness is a property of the version itself, and always needs to be updated at the same time the version is updated.

All of that said, happy to revert this change and go back to the drawing board, if you'd prefer to keep the logic in the agent and deal with clock skew a different way (e.g., @hugoShaka and I discussed the proxy also sending the current time).

Alternatively, if you'd like to keep the new model, I will:
- Move `agent_update_now` into `autoupdate_version`
- Find a way to better expose the "current version" concept via tctl

Reply:
Okay, that makes sense about moving a lot of this logic control plane side.

I still think we need to address the issue with critical updates. Like I mentioned in the previous comment, it's complex operationally. Once this issue is resolved, looks good to me.

Reply:
I think moving the "critical" setting into the `autoupdate_version` resource addresses this by tying the lifetime of the critical setting to the lifetime of the version. An admin would set the entire resource when the version changes.

For Cloud, this means that the Cloud platform declares a version to be critical, and it rolls out immediately (ignoring `agent_update_hour` in `cluster_maintenance_config`). When the next non-critical version is set, both the version and the "off" setting would be set atomically.

Let me know if this looks good:
https://github.com/gravitational/teleport/pull/40190/commits/88bdda43acdd684a490e512742ccd11292ccef69

Reply:
Flag for atomic setting: https://github.com/gravitational/teleport/pull/40190/commits/f98258cfa93cdb7c8b55d82a9e7c50cd5aa54ba8

(Might be worth a custom subcommand instead of set+flag? I can tweak later.)

Reply:
lgtm